components/brave_extension/extension/brave_extension/background/reducers/shieldsPanelReducer.ts

```diff
@@ -82,6 +83,13 @@ export default function shieldsPanelReducer (
       if (tab.active && tab.id) {
         state = shieldsPanelState.requestDataAndUpdateActiveTab(state, tab.windowId, tab.id)
         shieldsPanelState.updateShieldsIcon(state)
+        getViewPreferences()
+          .then((settings: chrome.braveShields.BraveShieldsViewPreferences) => {
+            if (state.persistentData.statsBadgeVisible === settings.statsBadgeVisible) {
+              state = shieldsPanelState.updatePersistentData(state, { statsBadgeVisible: settings.statsBadgeVisible })
```

Comment:
Setting state in a callback won't affect the state since the reducer will have already returned the state before this .then callback is run, e.g.
```js
console.log('1');
Promise.resolve().then(() => console.log('2'));
console.log('3');
 > 1
 > 3
 > 2
```
To solve all your issues with state, perhaps you should have a redux action called VIEW_PREFERENCES_CHANGED, and use `chrome.settingsPrivate.onPrefsChanged.addListener` to call this when the relevant items change.

Perhaps in `shieldsAPI.ts` you would have something line:
```ts
export function addViewPreferencesChangedListener(callback) {
  chrome.settingsPrivate.onPrefsChanged.addListener(changedPrefs => {
    // Reverse settings key / name lookup (can be done outside of this function)
    const settingsNameByKey = {}
    for (const name of Object.keys(settingsKeys)) {
      settingsNameByKey[settingsKeys[name].key] = { ...settingsKeys[name], name } 
    }
    // Get changed settings that are relevant
    const changedSettings: SetViewPreferencesData = {}
    for (const pref of changedPrefs) {  
      const setting = settingsNameByKey[pref.key]
      if (setting) {
        // Validate setting type
        if (pref.type !== setting.type) {
          throw new Error(`Unexpected settings type received for "${pref.key}". Expected: ${setting.type}, Received: ${pref.type}`)
        }
        // Valid
        changedSettings[setting.name] = pref.value
      }
    }
    if (Object.keys(changedSettings).length) {
      callback(changedSettings)
    }
  })
}
```

and then in your redux store setup you could use:
```js
shieldsAPI.addViewPreferencesChangedListener(actions.viewPreferencesChanged)
shieldsAPI.getViewPreferences.then(actions.viewPreferencesChanged)
```
and in the reducer store the value of these preferences in the state wherever you like.
wdyt?

Reply:
your proposal looks good but I think the call order doesn't matter since localStorage would update anyway, which is what is read by the reducer? also, shouldn't we avoid calling actions inside the store?

Reply:
I mean, I can't think of a case where the store would receive the wrong value as-is

Reply:
As-is, you are calling an async API inside the reducer, and then bypassing the action -> reducer flow to attempt to directly set the state inside that async callback. So that's the anti-pattern, but also the `state =` isn't actually doing anything here, so all this code can be removed, right? Or IMO better, is to move `shieldsAPI.addViewPreferencesChangedListener(actions.viewPreferencesChanged)` to your init function, whether that's the app init or store init (Setting up async events -> redux actions in the store init is a fine pattern IMO. The anti-pattern is calling actions inside the reducer function, which that wouldn't neccessarily be doing).

Reply:
e.g in background.ts you could setup the calls for the initial data which fires an action to set the initial data, as well as subscribe to events to update that data by firing an action again on change.